lib/Transformation.ts

```diff
@@ -156,25 +163,68 @@ function transformOperator(expr: Alg.OperatorExpression)
       throw new Err.UnknownOperator(expr.operator);
     }
     const op = expr.operator as C.RegularOperator;
-    const args = expr.args.map((a) => transformAlgebra(a));
+    const args = expr.args.map((a) => transformAlgebra(a, config));
     const func = regularFunctions.get(op);
     if (!hasCorrectArity(args, func.arity)) { throw new Err.InvalidArity(args, op); }
     return new E.Operator(args, func.apply);
   }
 }
 
+function wrapSyncFunction(f: SyncExtensionFunction, name: string): SimpleApplication {
+  return (args => {
+    let res;
+    try {
+      res = f(args.map(arg => arg.toRDF()));
+    }catch (e) {
+      throw new ExtensionFunctionError(name, e);
+    }
+    return transformRDFTermUnsafe(res);
+  });
+}
+
+function wrapAsyncFunction(f: AsyncExtensionFunction, name: string): AsyncExtensionApplication {
+  return (async args => {
+    let res;
+    try {
+      res = await f(args.map(arg => arg.toRDF()));
+    }catch (e) {
+      throw new ExtensionFunctionError(name, e);
+    }
+    return transformRDFTermUnsafe(res);
+  });
+}
+
 // TODO: Support passing functions to override default behaviour;
-export function transformNamed(expr: Alg.NamedExpression): E.NamedExpression {
+export function transformNamed(expr: Alg.NamedExpression, config: TransformAlgebraConfig)
```

Comment:
First, instead of passing the new type `TransformAlgebraConfig`, I would suggest passing `AsyncEvaluatorConfig | SyncEvaluatorConfig`.

Second, this functions becomes a bit too complex IMO.
This now assumes that the config can contain both a async and async callback.
I would suggest to keep these two cases distinct (and error early on if a value for both is given).

If you keep them separate, you can simplify this function quite a bit.
Furthermore, you can then make the return type of this function conditional: if `AsyncEvaluatorConfig` is given, return `E.AsyncExtensionExpression`; if `SyncEvaluatorConfig` is given, return `E.SyncExtensionExpression`.

Reply:
You're right, however, it has been mentioned in #16 that we might also want a custom implementation for functions. If we want to support this in the future isn't it good to have the argument object in place?

Reply:
> we might also want a custom implementation for functions
> If we want to support this in the future isn't it good to have the argument object in place?

We can still enable this when having `AsyncEvaluatorConfig | SyncEvaluatorConfig` as argument, no?

Reply:
No, it's probably cleaner to add a third argument than.

Reply:
I see only two, what third one do you have in mind?

Reply:
I meant when adding the custom function support.